pyrokinetics/diagnostics/Diagnostics.py

```diff
@@ -0,0 +1,265 @@
+                    kx=5,
+                    ky=5,
+                    s=1,
+                )
+                for theta in bxfft.theta
+            ]
+
+        # Main loop
+        x = xarray[np.newaxis, :]
+        y = yarray[:, np.newaxis]
+        points = np.empty((2, nturns, len(yarray), len(xarray)))
+
+        if use_invfft:
+            for iturn in range(nturns):
+                for ith in range(0, ntheta - 1, 2):
+                    dby = (
+                        self._invfft(ikxapar[:, :, ith], x, y, kx, ky)
+                        * bmag[ith]
+                        * fac2
+                    )
+                    dbx = (
+                        self._invfft(ikyapar[:, :, ith], x, y, kx, ky)
+                        * bmag[ith]
+                        * fac2
+                    )
+
+                    xmid = x + 2 * np.pi / ntheta * dbx * jacob[ith]
+                    ymid = y + 2 * np.pi / ntheta * dby * jacob[ith]
+
+                    dby = (
+                        self._invfft(ikxapar[:, :, ith + 1], xmid, ymid, kx, ky)
+                        * bmag[ith + 1]
+                        * fac2
+                    )
+                    dbx = (
+                        self._invfft(ikyapar[:, :, ith + 1], xmid, ymid, kx, ky)
+                        * bmag[ith + 1]
+                        * fac2
+                    )
+
+                    x = x + 4 * np.pi / ntheta * dbx * jacob[ith + 1]
+                    y = y + 4 * np.pi / ntheta * dby * jacob[ith + 1]
+
+                    y = np.where(y < ymin, ymax - (ymin - y), y)
+
+                    y = np.where(y > ymax, ymin + (y - ymax), y)
+
+                y = y + np.mod(fac1 * ((x - xmin) / Lx * dq + qmin), Ly)
+                y = np.where(y > ymax, ymin + (y - ymax), y)
+
+                points[0, iturn, :, :] = x
+                points[1, iturn, :, :] = y
+
+        else:
+            for iturn in range(nturns):
+                for ith in range(0, ntheta - 1, 2):
+                    dby = By[ith](x, y, grid=False) * bmag[ith] * fac2
+                    dbx = Bx[ith](x, y, grid=False) * bmag[ith] * fac2
+
+                    xmid = x + 2 * np.pi / ntheta * dbx * jacob[ith]
+                    ymid = y + 2 * np.pi / ntheta * dby * jacob[ith]
+
+                    dby = By[ith + 1](xmid, ymid, grid=False) * bmag[ith + 1] * fac2
+                    dbx = Bx[ith + 1](xmid, ymid, grid=False) * bmag[ith + 1] * fac2
+
+                    x = x + 4 * np.pi / ntheta * dbx * jacob[ith + 1]
+                    y = y + 4 * np.pi / ntheta * dby * jacob[ith + 1]
+
+                    y = np.where(y < ymin, ymax - (ymin - y), y)
+
+                    y = np.where(y > ymax, ymin + (y - ymax), y)
+
+                y = y + np.mod(fac1 * ((x - xmin) / Lx * dq + qmin), Ly)
+                y = np.where(y > ymax, ymin + (y - ymax), y)
+
+                points[0, iturn, :, :] = x
+                points[1, iturn, :, :] = y
```

Comment:
Can this be one for loop with the if statement around setting `dbx` and `dby`? Prevents repetition of code

Reply:
I think they are identical otherwise

Reply:
Yes, although I think the logic should remain outside the main loop for better performance.

Reply:
Is it possible to define functions to get dbx/dby outside of the loop and inside the branch and you can then use these functions inside the loop, avoiding the need for repetition?  I suppose there'll be some performance hit, but hopefully not much?

````
if use_invfft:
    def get_dbx(.....)
else 
    def get_dbx(....)

for .....
   for .....
````

Reply:
Yeah I see your point. Not sure how much of an impact it would have (would be roughly `2*nturns*ntheta` additional if statements to evaluate, though I'm not sure how much of an impact that has. I don't think we can vectorise it further as `x` and `y` are constantly changing.

Fine to leave as is.

Reply:
I performed various tests and there is little difference. I will push that in a moment.